### PR TITLE
New Pinecone Client

### PR DESCRIPTION
**Pull request by izellevy**
## Problem

Main branch is not supporting the new Pinecone client.

## Solution

Merging 0.4.0 release branch back into main.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [ ] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

Describe specific steps for validating this change.

### CODE REVIEW

File: .github/workflows/PR.yml [hunk 1]
```diff
@@ -83,11 +80,9 @@ jobs:
     - name: Run e2e tests
       id: e2e_tests
       if: github.event_name == 'merge_group'
-      continue-on-error: ${{ matrix.pinecone-plan == 'starter' }}
       env:
         INDEX_NAME: e2e-${{ steps.gen_suffix.outputs.INDEX_NAME_SUFFIX }}
-        PINECONE_ENVIRONMENT: ${{ matrix.pinecone-plan == 'paid' && secrets.PINECONE_ENVIRONMENT_3 || secrets.PINECONE_ENVIRONMENT_4 }}
-        PINECONE_API_KEY: ${{ matrix.pinecone-plan == 'paid' && secrets.PINECONE_API_KEY_3 || secrets.PINECONE_API_KEY_4 }}
+        PINECONE_API_KEY: ${{ secrets.PINECONE_API_KEY_5 }}
```
Comment (igiloh-pinecone) [inline .github/workflows/PR.yml:85]: Let's call it simply `PINECONE_API_KEY`?  
The `3` and `4` were used to indicate "pod" VS "serverless", not versions...

File: pyproject.toml [hunk 2]
```diff
@@ -33,11 +32,14 @@ pinecone-text = "^0.7.1"
 tokenizers = "^0.15.0"
 transformers = "^4.35.2"
 sentencepiece = "^0.1.99"
+pandas = "2.0.0"
+pyarrow = "^14.0.1"
```
Comment (igiloh-pinecone) [inline pyproject.toml:36]: Why do we need these? Are they coming from the client?  
Is there any code in Canopy that directly imports `pandas` or `pyarrow`? If not I think we can remove from here, and rely on the Client's dependencies indirectly.

Comment (igiloh-pinecone) [reply pyproject.toml:36]: I think these were actually coming from `pinecone-datasets` which we removed anyway. Not sure we need them anymore

File: src/canopy/knowledge_base/knowledge_base.py [hunk 3]
```diff
@@ -26,45 +22,46 @@
...
     """
+    pinecone_client = pinecone_client or _get_global_client()
+
+    try:
+        indexes = pinecone_client.list_indexes()
+    except Exception as e:
+        raise RuntimeError("Failed to connect to Pinecone. "
+                           "Please check your credentials and try again") from e
+
+    index_names = [_['name'] for _ in indexes.index_list['indexes']]
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:62]:
```suggestion
    index_names = indexes.names()
```
Apparently there's a nice method for this

File: src/canopy/knowledge_base/knowledge_base.py [hunk 4]
```diff
@@ -180,30 +179,35 @@ def __init__(self,
...
 
-    def _connect_index(self,
-                       connect_pinecone: bool = True
-                       ) -> None:
-        if connect_pinecone:
-            connect_to_pinecone()
+        self._default_spec = ServerlessSpec(
+            cloud="aws",
+            region="us-west-2"
+        )
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:199]: Why do we need to save this as `self.` member variable?  
You can just create it on the fly in `create_canopy_index()` if `spec is None`, can't you?

File: src/canopy/knowledge_base/knowledge_base.py [hunk 5]
```diff
@@ -668,17 +639,13 @@ def from_config(cls,
...
         return kb
 
-    @staticmethod
-    def _is_starter_env():
-        starter_env_suffixes = ("starter", "stage-gcp-0")
-        return os.getenv("PINECONE_ENVIRONMENT").lower().endswith(starter_env_suffixes)
+    @lru_cache(maxsize=1)
+    def _is_serverless_env(self):
+        description = self._pinecone_client.describe_index(self.index_name)
+        return "serverless" in description["spec"]
```
Comment (igiloh-pinecone) [inline src/canopy/knowledge_base/knowledge_base.py:649]: @izellevy `KnowledgeBase` has a dedicated `from_config()` constructor, which does some stuff with `index_params`.  That part should be removed.  

(In the near future we should probably remove this method altogether, but that's a different story..)

**Review by igiloh-pinecone — APPROVED** [review]
See a few minor comments

File: pyproject.toml (hunk 2 above)
Comment (izellevy) [reply pyproject.toml:36]: We need both the dependencies. It should have been a requirement in the first place. So after removing datasets it started to shout so I added them.

File: src/canopy/knowledge_base/knowledge_base.py (hunk 5 above)
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:649]: Yeah I removed that part.

File: src/canopy/knowledge_base/knowledge_base.py (hunk 4 above)
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:199]: I can actually let me do it

Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:199]: Done

File: src/canopy/knowledge_base/knowledge_base.py (hunk 3 above)
Comment (izellevy) [reply src/canopy/knowledge_base/knowledge_base.py:62]: Nice they added recently I guess will change to it

File: .github/workflows/PR.yml (hunk 1 above)
Comment (izellevy) [reply .github/workflows/PR.yml]: DONE

**Pull request merged**